### ISSUE TITLE
Bug: [no-unused-vars] doesn't report unused variables in module and ambient module declarations in definition files

### ISSUE THREAD

**Issue by ronami** [typescript-eslint/typescript-eslint#10713]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.3&fileType=.d.ts&code=KYDwDg9gTgLgBAOwIYFtgGcxIMbDgMQgjgG8AoOOAeirimElmABM5tp7sYAbATwH4KcGLzB4AQkihwAvHACMAbiGhG8EnAC%2ByzWTLNg2blLwoIzAK7c8AcgBGUm6SE046ABYQrrO3npqWQUoRMThJaTklFXBodS0dPVcPL24fPwZYwLIQiSlZBWVVWNJ4siA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tidmUQAmtAG4BDaKgyRE0aB2iRwYAL4hlQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
export namespace Foo {
  // reported correctly?
  type Bar = 1;
  export { };
}

declare module 'bar' {
  // should be reported?
  type Bar = 1;
  export { };
}

// should be reported?
type Bar = 1;
export { };
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unused-vars": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected the rule to report on the `bar` variable as unused, similar to TypeScript's report.

### Actual Result

The variable `bar` isn't marked as unused, though to my understanding it should be.

### Additional Info

Note that this shares some of the behavior described in #123.
